File: ode/pendulum_on_tube/julia/figure.jl
using Roots
using Plots

include("pot-config.jl")

function energy(u, rho)
    0.5*(1 + rho*u[1])^2*u[2]^2 + rho*sin(u[1]) + 1 - (1 + rho*u[1])*cos(u[1])
end

e0 = energy(u0, rho)
println("energy = ", e0)
amplitude = theta -> e0 - energy([theta, 0], rho)
theta_left = find_zero(amplitude, (-2*pi/3, 0))
theta_right = find_zero(amplitude, (2*pi/3, 0))
println("theta_left = ", theta_left)
println("theta_right = ", theta_right)

t = LinRange(0, 2*pi, 360)
theta = LinRange(theta_left, theta_right, N)

gr(size = (300, 300))
Plots.scalefontsizes()
Plots.scalefontsizes(0.4)

plotFigure = plot(
    rho*sin.(t),
    rho*cos.(t),
    seriestype = [:shape],
    linewidth = 0.7,
    color = :slateblue,
    aspect_ratio = 1,
    xlabel = "x",
    ylabel = "y",
    plot_title = "Pendulum on the tube",
    legend = false,
    fillalpha = 0.3,
    dpi = 300
)

plot!(
    plotFigure,
    [rho, rho],
    [0, -1],
    linewidth = 0.7,
    linecolor = :black
)

plot!(
    plotFigure,
    rho .+ 0.02*sin.(t),
    -1 .+ 0.02*cos.(t),
    seriestype = [:shape],
    linewidth = 1,
    color = :black,
)

plot!(
    plotFigure,
    map(x -> rho*cos(x) + (1 + rho*x)*sin(x), theta),
    map(x -> rho*sin(x) - (1 + rho*x)*cos(x), theta),
    linewidth = 1,
    linestyle = :dot,
    linecolor = :grey,
    alpha = 0.4
)

x_plot_size = xlims(plotFigure)[2] - xlims(plotFigure)[1]
y_plot_size = ylims(plotFigure)[2] - ylims(plotFigure)[1]
annotate!(
    xlims(plotFigure)[1] + 0.18*x_plot_size,
    ylims(plotFigure)[2] - 0.24*y_plot_size,
    text("\$\\mathcal{E}_0 = $(round(e0, digits=4))\$", :crimson, :right, 4)
)
annotate!(
    xlims(plotFigure)[1] + 0.18*x_plot_size,
    ylims(plotFigure)[2] - 0.76*y_plot_size,
    text("\$\\rho = $(round(rho, digits=4))\$", :darkslateblue, :right, 4)
)

gui(), readline()
savefig(plotFigure, "figure.png")

File: ode/pendulum_on_tube/python/figure.py
import numpy as np
from scipy.optimize import bisect
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

import potconfig as config

def energy(x):
    theta, omega = x
    rho = config.rho
    return 0.5*((1 + rho*theta)**2)*omega**2 + rho*np.sin(theta) + 1 - (1 + rho*theta)*np.cos(theta)

e0 = energy(config.x_init)
print("E0 = ", e0)
amplitude = lambda theta: e0 - energy([theta, 0])
theta_01 = bisect(amplitude, 0, np.pi/2)
theta_02 = bisect(amplitude, -np.pi/2, 0)
print("theta_01 : theta_02 = ", theta_01, " : ", theta_02)

circle1 = plt.Circle((0, 0), config.rho, color='mediumaquamarine')
plt.gca().add_patch(circle1)

arc1 = mpatches.Arc(
    (0, 0),
    2*config.rho,
    2*config.rho,
    angle=0,
    theta1=0,
    theta2=90,
    color='darkslateblue',
    linewidth=2
)
plt.gca().add_patch(arc1)

t = np.linspace(theta_01, theta_02, np.floor_divide(config.N_points, 2))
plt.plot(
    config.rho*np.cos(t) + (1 + config.rho*t)*np.sin(t), 
    config.rho*np.sin(t) - (1 + config.rho*t)*np.cos(t),
    color='grey',
    linestyle='dashed',
    linewidth=2,
    alpha=0.5
)

line1 = plt.plot(
    [config.rho, config.rho],
    [0, -1],
    color='darkslateblue',
    linewidth=2
)

circle2 = plt.Circle((config.rho, -1), 0.03, color='darkslateblue')
plt.gca().add_patch(circle2)

plt.axis("equal")
plt.title("Pendulum on the tube")
plt.xlabel(r"$x$")
plt.ylabel(r"$y$")

xmin, xmax, ymin, ymax = plt.axis()
plt.text(
    xmin + 0.08*(xmax - xmin), ymax - 0.08*(ymax - ymin),
    r"$\rho$ = {:.2f}".format(config.rho)
)
plt.text(
    xmin + 0.08*(xmax - xmin), ymin + 0.08*(ymax - ymin),
    r"$E_0$ = {:.4f}".format(e0)
)

plt.show()
